rr_common/src/planner/hill_climb_planner.cpp

```diff
@@ -0,0 +1,148 @@
+
+#include <parameter_assertions/assertions.h>
+
+namespace rr {
+
+template <class T>
+assertions::Assertion<T> container_not_empty{ [](const T& v) { return !v.empty(); }, "container not empty" };
+
+HillClimbPlanner::HillClimbPlanner(const ros::NodeHandle& nh, const rr::DistanceChecker& distanceChecker,
+                                   const rr::BicycleModel& bicycleModel)
+      : distance_checker_(distanceChecker), model_(bicycleModel), rand_gen_(), normal_distribution_(0, 1) {
+    assertions::getParam(nh, "segment_sections", segment_sections_, { container_not_empty<std::vector<int>> });
+    assertions::getParam(nh, "k_dist", k_dist_, { assertions::greater_eq(0.0) });
+    assertions::getParam(nh, "k_speed", k_speed_, { assertions::greater_eq(0.0) });
+    assertions::getParam(nh, "k_angle", k_angle_, { assertions::greater_eq(0.0) });
+    assertions::getParam(nh, "k_steering", k_steering_, { assertions::greater_eq(0.0) });
+    assertions::getParam(nh, "collision_penalty", collision_penalty_, { assertions::greater_eq(0.0) });
+    assertions::getParam(nh, "max_steering", max_steering_, { assertions::greater(0.0) });
+    assertions::getParam(nh, "num_workers", num_workers_, { assertions::greater(0) });
+    assertions::getParam(nh, "num_restarts", num_restarts_, { assertions::greater(0) });
+    assertions::getParam(nh, "neighbor_stddev", neighbor_stddev_, { assertions::greater(0.0) });
+    assertions::getParam(nh, "local_optimum_tries", local_optimum_tries_, { assertions::greater(0) });
+}
+
+void HillClimbPlanner::FillObstacleCosts(PlannedPath& plan) const {
+    if (plan.dists.size() != plan.path.size()) {
+        plan.dists.resize(plan.path.size());
+    }
+
+    plan.cost = 0;
+    plan.has_collision = false;
+    for (size_t i = 0; i < plan.path.size(); ++i) {
+        auto opt_dist = distance_checker_.GetCollisionDistance(plan.path[i].pose);
+        if (opt_dist) {
+            plan.cost += k_dist_ * std::exp(-std::max(0.01, opt_dist.value()));
+            plan.cost -= k_speed_ * plan.path[i].speed;
+            plan.cost += k_steering_ * plan.path[i].steer;
+            plan.cost += std::abs(plan.path[i].pose.theta) * k_angle_;
+            plan.dists[i] = opt_dist.value();
+        } else {
+            plan.cost += collision_penalty_ * (plan.path.size() - i);
+            std::fill(plan.dists.begin() + i, plan.dists.end(), 0);
+            plan.has_collision = true;
+            break;
+        }
+    }
+}
+
+void HillClimbPlanner::JitterControls(std::vector<double>& ctrl, double stddev) {
+    for (double& x : ctrl) {
+        x += normal_distribution_(rand_gen_) * stddev;
+        if (x < -max_steering_) {
+            x = -max_steering_;
+        } else if (x > max_steering_) {
+            x = max_steering_;
+        }
+    }
+}
+
+PlannedPath HillClimbPlanner::Plan(const rr::PCLMap& map) {
+    distance_checker_.SetMap(map);
+
+    auto descend_hill = [this, &map](PlannedPath& plan) {
```

Comment:
Can't you just do `[&]`?

Reply:
I think we have to capture `this` by value